velox/functions/lib/Re2Functions.cpp

```diff
@@ -809,71 +893,81 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
+  auto numAnyCharacterWildcards = 0;
+  auto numSingleCharacterWildcards = 0;
+  auto fixedPatternStart = 0;
+  auto numFixedPatterns = 0;
+  std::vector<std::string> fixedPatterns;
 
   while (i < patternLength) {
-    if (patternStr[i] == '%' || patternStr[i] == '_') {
-      // Ensures that pattern has a single contiguous stream of wildcard
-      // characters.
-      if (wildcardStart != -1) {
-        return std::make_pair(PatternKind::kGeneric, 0);
-      }
+    if (patternString[i] == '%' || patternString[i] == '_') {
       // Look till the last contiguous wildcard character, starting from this
       // index, is found, or the end of pattern is reached.
       wildcardStart = i;
       while (i < patternLength &&
-             (patternStr[i] == '%' || patternStr[i] == '_')) {
-        singleCharacterWildcardCount += (patternStr[i] == '_');
-        anyCharacterWildcardCount += (patternStr[i] == '%');
+             (patternString[i] == '%' || patternString[i] == '_')) {
+        numSingleCharacterWildcards += (patternString[i] == '_');
+        numAnyCharacterWildcards += (patternString[i] == '%');
         i++;
       }
+      numAnyWildcardStream += (numAnyCharacterWildcards > 0);
     } else {
-      // Ensure that pattern has a single fixed pattern.
-      if (fixedPatternStart != -1) {
-        return std::make_pair(PatternKind::kGeneric, 0);
-      }
       // Look till the end of fixed pattern, starting from this index, is found,
       // or the end of pattern is reached.
       fixedPatternStart = i;
       while (i < patternLength &&
-             (patternStr[i] != '%' && patternStr[i] != '_')) {
+             (patternString[i] != '%' && patternString[i] != '_')) {
         i++;
       }
+      numFixedPatterns++;
+      fixedPatterns.emplace_back(
+          patternString.substr(fixedPatternStart, i - fixedPatternStart));
     }
   }
+  numWildcards = numSingleCharacterWildcards + numAnyCharacterWildcards;
 
-  // Pattern contains wildcard characters only.
-  if (fixedPatternStart == -1) {
-    if (!anyCharacterWildcardCount) {
-      return {PatternKind::kExactlyN, singleCharacterWildcardCount};
+  // Pattern contains only wildcard characters.
+  if (!numFixedPatterns && numWildcards) {
+    if (!numAnyCharacterWildcards) {
+      return PatternMetadata(
+          PatternKind::kExactlyN, numSingleCharacterWildcards, 0, 0);
     }
-    return {PatternKind::kAtLeastN, singleCharacterWildcardCount};
-  }
-  // Pattern contains no wildcard characters (is a fixed pattern).
-  if (wildcardStart == -1) {
-    return {PatternKind::kFixed, patternLength};
+    return PatternMetadata(
+        PatternKind::kAtLeastN, numSingleCharacterWildcards, 0, 0);
   }
   // Pattern is generic if it has '_' wildcard characters and a fixed pattern.
-  if (singleCharacterWildcardCount) {
-    return {PatternKind::kGeneric, 0};
+  if (numFixedPatterns && numSingleCharacterWildcards) {
+    return PatternMetadata(PatternKind::kGeneric, 0, 0, 0);
```

Comment:
Move this generic case right to the end of these ifs

Reply:
Thanks for the suggestion, after updating the checks, this generic case is not really needed so removed it.